src/phase-2/scWETHv2.sol

```diff
@@ -0,0 +1,533 @@
+
+        uint256 totalFlashLoanAmount;
+        for (uint256 i; i < supplyBorrowParams.length; i++) {
+            totalFlashLoanAmount += supplyBorrowParams[i].borrowAmount;
+        }
+
+        scWETHv2.RebalanceParams memory params = scWETHv2.RebalanceParams({
+            repayWithdrawParams: new scWETHv2.RepayWithdrawParam[](0),
+            supplyBorrowParams: supplyBorrowParams,
+            wstEthToWethSwapData: "",
+            wethToWstEthSwapData: wethToWstEthSwapData
+        });
+
+        // needed otherwise counted as profit during harvest
+        totalInvested += totalInvestAmount;
+
+        _flashLoan(totalFlashLoanAmount, params);
+
+        emit Invested(totalInvestAmount, supplyBorrowParams);
+    }
+
+    /// @notice disinvest from lending markets in case of a loss
+    /// @param repayWithdrawParams : protocols to disinvest from and their respective amounts
+    function disinvest(RepayWithdrawParam[] calldata repayWithdrawParams, bytes calldata wstEthToWethSwapData)
+        external
+    {
+        onlyKeeper();
+        uint256 totalFlashLoanAmount;
+        for (uint256 i; i < repayWithdrawParams.length; i++) {
+            totalFlashLoanAmount += repayWithdrawParams[i].repayAmount;
+        }
+
+        RebalanceParams memory params = RebalanceParams({
+            repayWithdrawParams: repayWithdrawParams,
+            supplyBorrowParams: new SupplyBorrowParam[](0),
+            wstEthToWethSwapData: wstEthToWethSwapData,
+            wethToWstEthSwapData: ""
+        });
+
+        // take flashloan
+        _flashLoan(totalFlashLoanAmount, params);
+
+        emit DisInvested(repayWithdrawParams);
+    }
+
+    /// @notice reallocate funds between protocols (without any slippage)
+    // @param wstEthSwapAmount: amount of wstEth to swap to weth
+    function reallocate(
+        RepayWithdrawParam[] calldata from,
+        SupplyBorrowParam[] calldata to,
+        bytes calldata wstEthToWethSwapData,
+        bytes calldata wethToWstEthSwapData
+    ) external {
+        onlyKeeper();
+        uint256 totalFlashLoanAmount;
+        for (uint256 i; i < from.length; i++) {
+            totalFlashLoanAmount += from[i].repayAmount;
+        }
+
+        RebalanceParams memory params = RebalanceParams({
+            repayWithdrawParams: from,
+            supplyBorrowParams: to,
+            wstEthToWethSwapData: wstEthToWethSwapData,
+            wethToWstEthSwapData: wethToWstEthSwapData
+        });
+
+        // take flashloan
+        _flashLoan(totalFlashLoanAmount, params);
+
+        emit Reallocated(from, to);
+    }
+
+    //////////////////// VIEW METHODS //////////////////////////
+
+    function isSupported(uint256 _adapterId) public view returns (bool) {
+        return protocolAdapters.contains(_adapterId);
+    }
+
+    function getAdapter(uint256 _adapterId) external view returns (address adapter) {
+        (, adapter) = protocolAdapters.tryGet(_adapterId);
+    }
```

Comment:
do we really need this? if this is intended for the backend I don't think it's necessary since the backend can track the addresses of the adapter contracts on its own.